src/proxy_lib/proxy_lib.c

```diff
@@ -0,0 +1,253 @@
+ * that have to be called at the end.
+ * It uses 892 kB on Linux.
+ *
+ * Ad 2)
+ * The UMF pool allocator (the main one) is used from the creation to the destruction
+ * of a UMF pool to allocate memory needed by an application. It should be freed
+ * by an application.
+ */
+
+static UTIL_ONCE_FLAG Base_alloc_leak_initialized = UTIL_ONCE_FLAG_INIT;
+static umf_ba_linear_pool_t *Base_alloc_leak = NULL;
+static umf_memory_provider_handle_t OS_memory_provider = NULL;
+static umf_memory_pool_handle_t Proxy_pool = NULL;
+
+// it protects us from recursion in umfPool*()
+static __TLS int was_called_from_umfPool = 0;
+
+/*****************************************************************************/
+/*** The constructor and destructor of the proxy library *********************/
+/*****************************************************************************/
+
+void proxy_lib_create_common(void) {
+    umf_os_memory_provider_params_t os_params =
+        umfOsMemoryProviderParamsDefault();
+    enum umf_result_t umf_result;
+
+    umf_result = umfMemoryProviderCreate(umfOsMemoryProviderOps(), &os_params,
+                                         &OS_memory_provider);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "error: creating OS memory provider failed\n");
+        exit(-1);
+    }
+
+    umf_result = umfPoolCreate(umfJemallocPoolOps(), OS_memory_provider, NULL,
+                               0, &Proxy_pool);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "error: creating jemalloc pool manager failed\n");
+        exit(-1);
+    }
+    // The UMF pool has just been created (Proxy_pool != NULL). Stop using
+    // the linear allocator and start using the UMF pool allocator from now on.
+}
+
+void proxy_lib_destroy_common(void) {
+    umf_memory_pool_handle_t pool = Proxy_pool;
+    Proxy_pool = NULL;
+    umfPoolDestroy(pool);
+    umfMemoryProviderDestroy(OS_memory_provider);
+    // the "LEAK" linear base allocator is not destroyed at all
+}
+
+/*****************************************************************************/
+/*** Generic version of realloc() of linear base allocator *******************/
+/*****************************************************************************/
+
+static inline void *ba_generic_realloc(umf_ba_linear_pool_t *pool, void *ptr,
+                                       size_t size) {
+    if (size == 0) {
+        // it means free(ptr), but linear base allocator does not implement free()
+        return NULL;
+    }
+
+    assert(ptr); // it is verified in the main realloc()
+
+    size_t max_size = umf_ba_linear_pool_contains_pointer(pool, ptr);
+    assert(max_size > 0); // assert that pool contains the pointer
+
+    void *new_ptr = umf_ba_linear_alloc(pool, size);
+    if (!new_ptr) {
+        return NULL;
+    }
+
+    if (size > max_size) {
+        size = max_size;
+    }
+
+    memcpy(new_ptr, ptr, size);
```

Comment:
I don;t know if this is a big problem but you shouldn't call `memcpy(new_ptr, ptr, size);` - you should call memcpy with OLD size, not the new one I think. As it is now, you are potentially reading memory that belong to another allocation which can lead to races (and possibly be UB).

Reply:
1) We do NOT know the OLD size (storing it would complicate the allocator too much)
2) Reading the extra memory is not a problem, because we do not use it.

Reply:
Ok, let's keep it as is then. Just FYI, data races are always UB so even if we don't do anything with the data it might, in theory, lead to some problems (although in this scenario it's probably very unlikely).